pkg/kubelet/kuberuntime/kuberuntime_gc.go

```diff
@@ -0,0 +1,236 @@
+				containers: make([]*containerGCInfo, 0),
+				sandboxIDs: []string{s.GetId()},
+			}
+		}
+
+		containerStatuses, err := m.getKubeletContainerStatuses(s.GetId())
+		if err != nil {
+			return nil, err
+		}
+		for _, c := range containerStatuses {
+			evictablePods[podUID].containers = append(evictablePods[podUID].containers,
+				toContainerGCInfo(podUID, s.GetId(), c))
+		}
+
+		sort.Sort(containerByCreated(evictablePods[podUID].containers))
+	}
+
+	return evictablePods, nil
+}
+
+// enforceMaxPerPodContainer enforces max number of dead containers of each pod and
+// returns the number of total evictable containers.
+func enforceMaxPerPodContainer(evictablePods map[string]*podGCInfo, maxPerPodContainer int) int {
+	totalEvictableContainers := 0
+	for _, pod := range evictablePods {
+		if maxPerPodContainer > 0 {
+			pod.removeOldestNContainers(pod.containersNum() - maxPerPodContainer)
+		}
+		totalEvictableContainers += pod.containersNum()
+	}
+
+	return totalEvictableContainers
+}
+
+// GarbageCollect removes dead containers using the specified container gc policy.
+// It consists of following steps:
+// * gets evictable pods with belonging containers and sandboxes which are not active and created more than gcPolicy.MinAge ago.
+// * removes oldest dead containers for each pod by enforcing gcPolicy.MaxPerPodContainer.
+// * removes oldest dead containers by enforcing gcPolicy.MaxContainers.
+// * removes dead sandboxes with zero containers.
+func (m *kubeGenericRuntimeManager) GarbageCollect(gcPolicy kubecontainer.ContainerGCPolicy, allSourcesReady bool) error {
+	evictablePods, err := m.getEvictablePods(gcPolicy.MinAge)
+	if err != nil {
+		glog.Warningf("getEvictablePods failed: %v", err)
+		return err
+	}
+
+	glog.V(4).Infof("GarbageCollect gets pods %q for evicting", evictablePods)
+
+	// Enforce max number of dead containers of each pod
+	totalEvictableContainers := enforceMaxPerPodContainer(evictablePods, gcPolicy.MaxPerPodContainer)
+	// Enforce max number of containers
+	if gcPolicy.MaxContainers > 0 && totalEvictableContainers > gcPolicy.MaxContainers {
+		// sort all evitable containers by createdAt (newest first)
+		allEvictableContainers := make([]*containerGCInfo, 0)
+		for _, pod := range evictablePods {
+			allEvictableContainers = append(allEvictableContainers, pod.containers...)
+		}
+		sort.Sort(containerByCreated(allEvictableContainers))
+
+		// Remove oldest (totalEvictableContainers - gcPolicy.MaxContainers) containers
+		for i := gcPolicy.MaxContainers; i < totalEvictableContainers; i++ {
+			container := allEvictableContainers[i]
+			containerList := evictablePods[container.podUID].containers
+			// remove the container from runtime service and also update podGCInfo.containers
+			containerIndex := evictablePods[container.podUID].removeContainer(container.containerID)
+			evictablePods[container.podUID].containers = append(containerList[:containerIndex], containerList[containerIndex+1:]...)
```

Comment:
The three steps required three different eviction units:
 1. MaxPerPodContainer: `(pod UID, container name)`. Grouping by pods seems ok.
 2. MaxContainers: `(container name)`. Grouping by pods doesn't make sense.
 3. sandbox cleanup: `(sandbox id)`. Grouping by pods seems ok. 

If we can't change the order, there are some necessary transitions in between.
I suggest that we do these in the order of (1), (3), (2):
  * Both (1) and (3) can group by pods. At the end, we can convert them into a list of containers and do (2). No more funky business of deleting containers in the evictable pods.
  * Before trying to delete oldest containers globally, i.e., (2), we can make sure all the redundant sandboxes are cleaned up first.
  * The downside is that we won't be able to remove redundant sandboxes created by (2) until the next GC cycle, but I think that's fine for now.

WDYT? Does this make sense?